lib/strings.nix

```
@@ -754,15 +754,27 @@ rec {
        => 1337
        toInt "-4"
        => -4
+       toInt " 123 "
+       => 123
+       toInt "00024"
+       => 24
        toInt "3.14"
        => error: floating point JSON numbers are not supported
   */
   # Obviously, it is a bit hacky to use fromJSON this way.
   toInt = str:
-    let may_be_int = fromJSON str; in
-    if isInt may_be_int
-    then may_be_int
-    else throw "Could not convert ${str} to int.";
+    let
+      strippedInput = match "[[:space:]]*(0*)(.*)" str;
+      isNonZeroEmpty = match "[[:space:]]*" (lib.last strippedInput) == [];
+      isZeroNonEmpty = head strippedInput != "";
+      mayBeInt = fromJSON (lib.last strippedInput);
+    in
+      if isNonZeroEmpty && isZeroNonEmpty
+      then 0
+      else
+      if isInt mayBeInt
```

```suggestion
      strippedInput = match "[[:space:]]*0*([[:digit:]]+)[[:space:]]*" str;
      isZero = (match "[[:space:]]*0+[[:space:]]*" str) == [];
      mayBeInt = fromJSON (lib.head strippedInput);
    in
      if isZero
      then 0
      else
      if strippedInput != null && isInt mayBeInt
```

I find this form makes it easier to read for someone who may not even be familiar with regex.
Since we do not support floating point JSON numbers, it is safe to just match against digit.

Also, ... I'm not totally sure if this function should strip spaces.

---

I find the proposed regex to be somewhat harder to read (comparatively it's noisier and harder to reason about). I suspect if we want to make the variant in the PR easier to read, it'd be preferable to just leave a comment above it as a hint to the reader. Example: 

```
      # Match leading whitespace, capture zero padding if present in group 0, 
      # and capture the rest of the input in group 1. 
      strippedInput = match "[[:space:]]*(0*)(.*)" str;
```

The reason I matched against `.*` instead of digits is because that allows `fromJSON` to do the error handling instead which changes the behavior of `toInt` as little as necessary.

Additionally I think it'd be preferable to do as few regex match operations as possible to reduce the potential performance overhead of adding this change to an otherwise fairly common operation.

If we want to make the code surrounding the match a little easier, it may be more intuitive to use `elemAt` instead of `head` and `last`.

---

As for the spaces being stripped, thats because that's the behavior that the function currently has. Spaces are only handled in the regex because there's not a good way of continuing to support conversion of strings with spaces other than this.

---

> I suspect if we want to make the variant in the PR easier to read, it'd be preferable to just leave a comment above it as a hint to the reader

I like this (regex is always daunting). The approach you take is a matter of preference. Still, I think the test for zero should be simplified as a single regex test instead of chaining the result

> The reason I matched against .* instead of digits is because that allows fromJSON to do the error handling instead which changes the behavior of toInt as little as necessary.

That makes sense. The only thing though is that you need to be careful with how you strip the zeros. For example, as it currently is, the string "000     123" gets parsed as 123 (which I think is undesirable).

> If we want to make the code surrounding the match a little easier, it may be more intuitive to use elemAt instead of head and last.

I like this.

> As for the spaces being stripped, thats because that's the behavior that the function currently has. 

Sorry, I didnt know that.

---

> That makes sense. The only thing though is that you need to be careful with how you strip the zeros. For example, as it currently is, the string "000 123" gets parsed as 123 (which I think is undesirable).

Ah yep I forgot to check for that. I'll add that in

---

after further consideration I'll probably use your suggested change to the regex with the other changes we've discussed.